qiskit/primitives/containers/data_bin.py

```diff
@@ -15,78 +15,78 @@
 """
 from __future__ import annotations
 
-from collections.abc import Iterable
-from dataclasses import make_dataclass
+from typing import Any, Iterable
 
 
 class DataBinMeta(type):
-    """Metaclass for :class:`DataBin` that adds the shape to the type name.
+    """This metaclass no longer has a purpose and will be removed."""
 
-    This is so that the class has a custom repr with DataBin<*shape> notation.
-    """
 
-    def __repr__(cls):
-        name = cls.__name__
-        if cls._SHAPE is None:
-            return name
-        shape = ",".join(map(str, cls._SHAPE))
-        return f"{name}<{shape}>"
+class DataBin(metaclass=DataBinMeta):
+    """Namespace for storing data.
 
+    .. code-block:: python
 
-class DataBin(metaclass=DataBinMeta):
-    """Base class for data bin containers.
+        data = DataBin(
+            alpha=BitArray.from_bitstrings(["0010"]),
+            beta=np.array([1.2])
+        )
+
+        print("alpha data:", data.alpha)
+        print("beta data:", data.beta)
 
-    Subclasses are typically made via :class:`~make_data_bin`, which is a specialization of
-    :class:`make_dataclass`.
     """
 
     _RESTRICTED_NAMES = ("_RESTRICTED_NAMES", "_SHAPE", "_FIELDS", "_FIELD_TYPES")
-    _SHAPE: tuple[int, ...] | None = None
-    _FIELDS: tuple[str, ...] = ()
-    """The fields allowed in this data bin."""
-    _FIELD_TYPES: tuple[type, ...] = ()
-    """The types of each field."""
+
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+    def __setattr__(self, key, value):
+        raise NotImplementedError
 
     def __len__(self):
-        return len(self._FIELDS)
+        return len(self.__dict__)
 
     def __repr__(self):
         vals = (f"{name}={getattr(self, name)}" for name in self._FIELDS if hasattr(self, name))
-        return f"{type(self)}({', '.join(vals)})"
+        return f"{type(self).__name__}({', '.join(vals)})"
 
+    # The following properties exist to provide support to legacy private class attributes which
+    # gained widespread usage in several internal projects due to the lack of alternatives prior to
+    # qiskit 1.1. These properties will be removed once the internal projects have made the
+    # appropriate changes.
 
+    @property
+    def _FIELDS(self) -> tuple[str, ...]:  # pylint: disable=invalid-name
+        return tuple(self.__dict__)
+
+    @property
+    def _FIELD_TYPES(self) -> tuple[Any, ...]:  # pylint: disable=invalid-name
+        return tuple(type(val) for val in self.__dict__.values())
+
+    @property
+    def _SHAPE(self) -> tuple[int, ...] | None:  # pylint: disable=invalid-name
+        return None
+
+
+# pylint: disable=unused-argument
 def make_data_bin(
     fields: Iterable[tuple[str, type]], shape: tuple[int, ...] | None = None
 ) -> DataBinMeta:
```

Comment:
How about changing the return type to `type[DataBin]` as I did in #12129?
#12129 adds some methods to `DataBin`. But since they are not part of `DataBinMeta`, they are not recognized by IDE.